### PR TITLE
Functions codegen

### PR DESCRIPTION
**Pull request by bzsurbhi**
#### Description of changes
- Added functions codegen changes

**Properties getting added:**
export const amplifymigrateecd61fbc = defineFunction({
    entry: "./amplifymigrateecd61fbc/src/handler.ts",
    name: "amplifymigrateecd61fbc-dev",
    timeoutSeconds: 25,
    memoryMB: 128,
    environment: { ENV: "dev", REGION: "us-west-2" },
    runtime: 18
});

- Also, fixed some Auth codegen bugs related to mfa and userattributes props of auth

#### Description of how you validated changes
Deployed the changes to sandbox, verified function resource getting created successfully

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-migration/src/index.ts [hunk 1]
```diff
@@ -117,7 +131,10 @@ export async function execute() {
       () => getAuthTriggersConnections(),
     ),
     dataDefinitionFetcher: new DataDefinitionFetcher(backendEnvironmentResolver, amplifyStackParser),
+    functionsDefinitionFetcher: new AppFunctionsDefinitionFetcher(lambdaClient, backendEnvironmentResolver, stateManager),
     analytics: new AppAnalytics(appId),
     logger: new AppContextLogger(appId),
+    backendEnvironmentName: backendEnvironment?.environmentName,
```
Comment (abhi7cr) [inline packages/amplify-migration/src/index.ts:137]: can this be null?

File: packages/amplify-migration/src/app_functions_definition_fetcher.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    private backendEnvironmentResolver: BackendEnvironmentResolver,
+    private stateManager: StateManager,
+  ) {}
+
+  getDefinition = async (): Promise<FunctionDefinition[] | undefined> => {
+    const backendEnvironment = await this.backendEnvironmentResolver.selectBackendEnvironment();
+    assert(backendEnvironment?.stackName);
+
+    const meta = this.stateManager.getMeta();
+
```
Comment (abhi7cr) [inline packages/amplify-migration/src/app_functions_definition_fetcher.ts:24]: can we add an assert for `meta` to make sure is not nullable/undefined? We can then remove the `?` in the below statements.

File: packages/amplify-migration/src/index.ts (hunk 1 above)
Comment (bzsurbhi) [reply packages/amplify-migration/src/index.ts:137]: `backendEnvironment` can be undefined 

```
const backendEnvironment = await backendEnvironmentResolver.selectBackendEnvironment();

selectBackendEnvironment can return an undefined value
selectBackendEnvironment = async (): Promise<BackendEnvironment | undefined>
```

File: packages/amplify-migration/src/app_functions_definition_fetcher.ts (hunk 2 above)
Comment (bzsurbhi) [reply packages/amplify-migration/src/app_functions_definition_fetcher.ts:25]: added

**Review by abhi7cr — APPROVED** [review]

File: packages/amplify-gen1-codegen-auth-adapter/src/auth_render_adapter.ts [hunk 3]
```diff
@@ -16,6 +16,7 @@ import {
   LoginOptions,
   Scope,
 } from '@aws-amplify/amplify-gen2-codegen';
+import { AttributeMappingRule } from '@aws-amplify/amplify-gen2-codegen/src/auth/source_builder';
```
Comment (sobolk) [inline packages/amplify-gen1-codegen-auth-adapter/src/auth_render_adapter.ts:19]: Is it intentional that changes in `auth-adapter` are part of `function codegen` ?
I.e. this doesn't match PR description.

File: packages/amplify-gen1-codegen-function-adapter/API.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
+## API Report File for "@aws-amplify/amplify-gen1-codegen-auth-adapter"
```
Comment (sobolk) [inline packages/amplify-gen1-codegen-function-adapter/API.md:1]: why is `auth-adapter` report under `function-adapter` ?

File: packages/amplify-gen1-codegen-function-adapter/API.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+> Do not edit this file. It is a report generated by [API Extractor](https://api-extractor.com/).
+
+```ts
+
+import { AuthDefinition } from '@aws-amplify/amplify-gen2-codegen';
+import { LambdaConfigType } from '@aws-sdk/client-cognito-identity-provider';
+import { PasswordPolicyPath } from '@aws-amplify/amplify-gen2-codegen';
+import { UserPoolType } from '@aws-sdk/client-cognito-identity-provider';
```
Comment (sobolk) [inline packages/amplify-gen1-codegen-function-adapter/API.md:10]: are we using cognito types in functions ?

**Review by sobolk — COMMENTED** [review]
Could you please run `yarn cloud-pr` to trigger codebuild job and make sure it passes ?

File: packages/amplify-gen1-codegen-function-adapter/API.md (hunk 5 above)
Comment (bzsurbhi) [reply packages/amplify-gen1-codegen-function-adapter/API.md]: No, updated the API.md file

File: packages/amplify-gen1-codegen-function-adapter/API.md (hunk 4 above)
Comment (bzsurbhi) [reply packages/amplify-gen1-codegen-function-adapter/API.md]: Updated

File: packages/amplify-gen1-codegen-auth-adapter/src/auth_render_adapter.ts (hunk 3 above)
Comment (bzsurbhi) [reply packages/amplify-gen1-codegen-auth-adapter/src/auth_render_adapter.ts:19]: Yes, the changes are intentional. I updated the description

File: packages/amplify-gen2-codegen/API.md [hunk 6]
```diff
@@ -198,7 +231,11 @@ export interface StorageRenderParameters {
 export type StorageTriggerEvent = 'onDelete' | 'onUpload';
 
 // @public (undocumented)
-export type UserPoolMfaConfig = 'OFF' | 'ON' | 'OPTIONAL';
+export type UserPoolMfaConfig = 'OFF' | 'REQUIRED' | 'OPTIONAL';
+
+// Warnings were encountered during analysis:
+//
+// src/auth/source_builder.ts:101:3 - (ae-forgotten-export) The symbol "AttributeMappingRule" needs to be exported by the entry point index.d.ts
```
Comment (sobolk) [inline packages/amplify-gen2-codegen/API.md:238]: please fix this.

**Review by sobolk — APPROVED** [review]

**Pull request merged**